### PR TITLE
Proof of concept druid running on web

### PR DESCRIPTION
**Pull request by tedsta**
Hey Raph long time no chat!

I've been wanting to do some hobby web tools and since somebody already did the heavy lifting getting piet working on wasm I figured I'd give this a shot.

Is web something you are interested in targeting with Druid?

image

### CODE REVIEW

**Comment by raphlinus** [comment]
Hi @tedsta! Indeed, it's been a while.

Yes, targeting web is *definitely* something I'm interested in. I wanted to respond quickly before doing a full review. There is some foundational work in druid remaining that will affect ports, so it's a project management question whether we should limit the number of ports while that's churning, or bring all ports along. There's similarly a [Linux port](https://github.com/xi-editor/druid/compare/master...bobtwinkles:gtk) (using a gtk dependency) where we need to make some similar decisions.

I hope to review this in more detail soon; this week is challenging for finding chunks of time.

**Comment by tedsta** [comment]
Take your time - I should emphasize the proof-of-concept-ness of this PR a bit more. Also note this is the first time I've even looked at wasm and `web_sys` stuff.

I switched over to the calculator example and while it renders ok it doesn't react to button presses properly, so I'll be looking at that next.

I think it could be helpful to have the web port in the churn so that assumptions aren't locked in that aren't valid on web. But I understand if it's too much - I could go play with the linux port instead.

Current state (questions are just things I'm asking myself):
- Everything except rendering and mouse events is stubbed out
- Relies on the DOM already having a canvas element named "canvas". Should we just create a canvas in the body? Assume the body is empty at start?
- There is no main loop, just set up the callbacks. The rust entry point actually exits. This makes the "idle queue" a bit problematic... my code assumes that idle callbacks will only be queued after some kind of UI event, but I that isn't valid when external things like network messages can update the GUI.

I need to do some learning here - I don't know when callbacks I register with `web_sys` are called. I assumed they are called on the main thread, and so if I hog the thread with a main loop, they would never be called? (haven't actually tested) Or maybe the callbacks will preempt my main loop?

- `std::time` doesn't work on wasm32-unknown-unknown, so I made a little `Timer` object with an implementation for wasm32 and another for everyone else.

- How do you exit the web app? Close the tab? Do web APIs give a way to do that?

File: druid-shell/src/web/mod.rs [hunk 1]
```diff
@@ -166,7 +167,7 @@ fn setup_web_callbacks(window_state: &Rc<WindowState>) {
 
             let event = MouseEvent {
                 x: event.offset_x() as i32,
-                y: event.offset_x() as i32,
```
Comment (tedsta) [inline druid-shell/src/web/mod.rs]: .... figured out why half of the calc example buttons iddn't work... :roll_eyes:

**Comment by msand** [comment]
Hello! This seems like a interesting project. I'm thinking you probably want to use request_animation_frame for the main loop: https://github.com/rustwasm/wasm-bindgen/blob/fcee4656923a1128d0a2faf781242ee2d7835db4/examples/request-animation-frame/src/lib.rs#L10-L14

https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.Window.html#method.request_animation_frame
This should allow the js event loop to function performantly and not hog the main ui thread.

**Comment by tedsta** [comment]
Thanks for the tip @msand, idle callbacks now use `request_animation_frame(_)`

**Comment by msand** [comment]
@tedsta You're welcome 👍Thanks for working on this! I just realized, if it's actually idle you want rather than high fps render loop, then you might want to change to requestIdleCallback instead: https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.Window.html#method.request_idle_callback

**Comment by jjl** [comment]
This rebases cleanly against master and works (more or less).

I've pushed some changes to my [web-hidpi](https://github.com/jjl/druid/tree/web-hidpi) branch which is a continuation of your web branch:
- The DPI calculation caused 1/3 of the content in each dimension to be clipped, this is rectified
- Mouse positions are now translated into physical positions instead of css positions
- Added window resizing

**Comment by jjl** [comment]
I have been thinking about next steps.

At the minute the event loop is handled through the window. I would like to move this to the run loop but it's not immediately clear what the correct way of doing this is, particularly with regard to supporting other platforms - where the cross platform api stops and the per-platform begins is not entirely clear to me at present.I am thinking to copy what is done in windows for now, but will probably wait to see what happens for multiple window support on other platforms before looking for an alternative solution.

Speaking of windows... We should decide what constitutes a window on the web backend. My current thoughts are that yes, it should be a canvas but no, it should not necessarily be full screen. Perhaps full screen can be a mode opted into in the same way one might for other platforms.

This leaves open a few questions:

* Should we map windows to canvas elements such that when multiple windows are supported on other platforms, it maps multiple canvas elements rendered in one document? (my view: yes)
* Should these be optionally individually resizable? (i think so, because we'd want that on other platforms)
* Should creating and closing a window map to creating and destroying a canvas? (seems logical)
* What are we to do about ownership of windows and the run loop? (???)

**Comment by jjl** [comment]
Seems I may have spoken too soon, my master was out of date and the muggle merge has broken it. Perhaps going through it will answer some of my other questions

**Comment by jjl** [comment]
I seem to have everything except Text working in my local copy - a shame we need text everywhere for the calculator.

It's a bit awkward trying to infer what the web structure should be from the windows version, but I've gotten most of it down, I think. I'm currently just working out if i can get a mutable canvas context ref to the WinCtxOwner without requesting the context afresh.

**Comment by xStrom** [comment]
Now that #759 has landed, is there still a reason to keep this PR open?

**Comment by raphlinus** [comment]
Nope, this was the basis of that, so safe to close. Thanks for the heads-up!

**Pull request closed without merging**